gh-99991: improve docs on str.encode and bytes.decode (#100198)
python/cpython: Doc/library/stdtypes.rst

.. method:: str.encode(encoding="utf-8", errors="strict")

>    Return an encoded version of the string as a bytes object. Default encoding

Reviewer: ```suggestion
   Return the string encoded to :class:`bytes`.
```

* Following the original issue, keep the summary line in its own paragraph for clarity and readability.
* Per the original issue, properly markup/link the `bytes` object.
* Also, per the general principles of reference type documentation, keep this more concise.

   is ``'utf-8'``.  And for a list of possible encodings, see section

>    :ref:`standard-encodings`.

Reviewer: ```suggestion
   *encoding* defaults to ``'utf-8'``;
   see :ref:`standard-encodings` for possible values.
```

* Per the original issue, give the *encoding* argument it's own paragraph for readability
* Explicitly state and format the parameter name at the beginning per the docs style guide
* Fix and greatly improve grammar/phrasing
* Be more concise and austere in reference docs

   *errors* may be given to set a different error handling scheme.  The
   default for *errors* is ``'strict'``, meaning that encoding errors raise
   a :exc:`UnicodeError`. Other possible values are ``'ignore'``,
   ``'replace'``, ``'xmlcharrefreplace'``, ``'backslashreplace'`` and any
   other name registered via :func:`codecs.register_error`, see section

>    :ref:`error-handlers`.

Reviewer: ```suggestion
   *errors* controls how encoding errors are handled.
   If ``'strict'`` (the default), a :exc:`UnicodeError` exception is raised.
   Other possible values are ``'ignore'``,
   ``'replace'``, ``'xmlcharrefreplace'``, ``'backslashreplace'`` and any
   other name registered via :func:`codecs.register_error`.
   See :ref:`error-handlers` for details.
```

* Be more direct, descriptive and conventional with summary line
* Be more concise overall and avoid repetition
* Avoid grammar error (comma splice)

   By default, the *errors* argument is not checked for best performances, but

>    only used at the first encoding error. Enable the :ref:`Python Development

Reviewer: I was not able to actually make this as suggestion, but this implements the revision from the issue, while still including the key details from the previous version.

```rst
   For performance reasons, the value of *errors* is not checked for validity
   unless an encoding error actually occurs,
   :ref:`Python Development Mode <devmode>` is enabled
   or a :ref:`debug build <debug-build>` is used.
```

            bytearray.decode(encoding="utf-8", errors="strict")

>    Return a string decoded from the given bytes.  Default encoding is

Reviewer: ```suggestion
   Return the bytes decoded to a :class:`str`.
```

As above:
* Following the original issue, keep the summary line in its own paragraph for clarity and readability.
* Per the original issue, properly markup/link the `str` object.
* Also, per the general principles of reference type documentation, keep this more concise.

>    ``'utf-8'``.  And for a list of possible encodings, see section
>    :ref:`standard-encodings`.

Reviewer: ```suggestion
   *encoding* defaults to ``'utf-8'``;
   see :ref:`standard-encodings` for possible values.
```

As above:
* Per the original issue, give the *encoding* argument it's own paragraph for readability
* Explicitly state and format the parameter name at the beginning per the docs style guide
* Fix and greatly improve grammar/phrasing
* Be more concise and austere in reference docs

>    *errors* may be given to set a different error handling scheme.  The
>    default for *errors* is ``'strict'``, meaning that encoding errors raise a
>    :exc:`UnicodeError`.  Other possible values are ``'ignore'``, ``'replace'``
>    and any other name registered via :func:`codecs.register_error`, see
>    section :ref:`error-handlers`.

Reviewer: ```suggestion
   *errors* controls how decoding errors are handled.
   If ``'strict'`` (the default), a :exc:`UnicodeError` exception is raised.
   Other possible values are ``'ignore'``, ``'replace'``,
   and any other name registered via :func:`codecs.register_error`.
   See :ref:`error-handlers` for details.
```

* Fix "encoding errors" -> "decoding errors"

As above:
* Be more direct, descriptive and conventional with summary line
* Be more concise overall and avoid repetition
* Avoid grammar error (comma splice)

   By default, the *errors* argument is not checked for best performances, but

>    only used at the first decoding error. Enable the :ref:`Python Development

Reviewer: I was not able to actually make this as suggestion, but this implements the revision from the issue, while still including the key details from the previous version.

```rst
   For performance reasons, the value of *errors* is not checked for validity
   unless a decoding error actually occurs,
   :ref:`Python Development Mode <devmode>` is enabled
   or a :ref:`debug build <debug-build>` is used.
```
